Game/data/building_systems/heat_system/heat_system.gd

```diff
@@ -15,78 +22,70 @@ var heat_flow_graphs: Dictionary[int, HeatFlowGraph]
 ## weight from sink -> omni-sink: consumed heat flow from sources
 var steady_state_flows: Dictionary[int, HeatFlowGraph]
 
-## List of all heat sources.
-var heat_sources: Dictionary[int, Array]
-
-## List of all heat sinks.
-var heat_sinks: Dictionary[int, Array]
 
 func _ready() -> void:
-	ComponentManager.component_added.connect(_on_component_added)
-	ComponentManager.component_removed.connect(_on_component_removed)
-	Model.game_ready.connect(_on_game_ready)
-
-
-## When a heat component is added, update the internal state and re-calculate the steady state
-## flow of heat.
-func _on_component_added(component: BuildingComponent) -> void:
-	if component is HeatComponent:
-		var player_id: int = component.building_entity.player_id
+	ComponentManager.component_added.connect(on_component_added)
+	ComponentManager.component_removed.connect(on_component_removed)
+	Model.game_ready.connect(on_game_ready)
 
-		heat_flow_graphs[player_id].add_building(component)
 
-		if component.is_source:
-			heat_sources[player_id].append(component)
-		elif component.is_sink:
-			heat_sinks[player_id].append(component)
+## When a heat component is added, update the internal state.
+func on_component_added(component: BuildingComponent) -> void:
+	if component is not HeatComponent:
+		return
 
-		Model.add_heat_data_at.rpc(
-			component.building_entity.player_id,
-			component.building_entity.position,
-			component.heat,
-			component.heat_capacity)
+	var player_id: int = component.building_entity.player_id
 
-		calculate_steady_state_flow(player_id)
+	heat_flow_graphs_current[player_id].add_building(component)
+	heat_flow_graphs_dirty[player_id] = true
 
-		print_flow_rates()
+	Model.add_heat_data_at.rpc(
+		component.building_entity.player_id,
+		component.building_entity.position,
+		component.heat,
+		component.heat_capacity,
+		component.heat_state
+	)
 
 
-## When a heat component is removed, update the internal state and re-calculate the steady state
-## flow of heat.
-func _on_component_removed(component: BuildingComponent) -> void:
-	if component is HeatComponent:
-		var player_id: int = component.building_entity.player_id
+## When a heat component is removed, update the internal state.
+func on_component_removed(component: BuildingComponent) -> void:
+	if component is not HeatComponent:
+		return
 
-		if component.is_source:
-			heat_sources.erase(component)
-		elif component.is_sink:
-			heat_sinks.erase(component)
+	var player_id: int = component.building_entity.player_id
 
-		heat_flow_graphs[player_id].remove_building(component)
+	heat_flow_graphs_current[player_id].remove_building(component)
+	heat_flow_graphs_dirty[player_id] = true
 
-		Model.remove_heat_data_at.rpc(
-			player_id,
-			component.building_entity.position)
-
-		calculate_steady_state_flow(player_id)
-
-		print_flow_rates()
+	Model.remove_heat_data_at.rpc(player_id, component.building_entity.position)
```

Comment:
I think the decision of who to inform about the heat data should be made in the model, not in the heat system. Additionally, the *timing* of when to inform anyone about anything should be handled in the model, not here.